sdks/python/apache_beam/transforms/external.py

```diff
@@ -42,6 +44,63 @@
   grpc = None
 # pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
 
+DEFAULT_EXPANSION_SERVICE = 'localhost:8097'
+FILTERED_CODERS = {
+  'beam:coder:length_prefix:v1',
+}
+
+
+def iter_urns(coder, context=None):
+  yield coder.to_runner_api_parameter(context)[0]
+  for child in coder._get_component_coders():
+    for urn in iter_urns(child, context):
+      yield urn
+
+
+class External(ptransform.PTransform):
+  """
+  Base class for transforms that produce external transforms.
+  """
+  _urn = None
+
+  def __init__(self, expansion_service=None):
+    self.expansion_service = expansion_service or DEFAULT_EXPANSION_SERVICE
+
+  def get_config_args(self):
+    """
+    :return: dictionary of encoded configuration values to pass to
+             ExternalConfigurationPayload
+    """
+    raise NotImplementedError
+
+  @classmethod
+  def config_value(cls, obj, coder):
```

Comment:
In terms of ease of use, we could make coder an optional argument and infer it based on the type of obj. Even better, we could let  get_config_args return a dict whose values need not be pre-wrapped and just do the wrapping here.

Reply:
I agree, we can make this even easier, and we should.  This suggestion got me thinking: a lightweight schema class could be a nice way to do this.  Here's an example using [`attrs`](https://github.com/python-attrs/attrs):

```python
@attrs
ReadFromKafkaSchema(object):
    consumer_config = attrib(type=Iterable[Tuple[str, str]])
    topics = attrib(type=Iterable[str])
    key_deserializer = attrib(type=str)
    value_deserializer = attrib(type=str)
    
class ReadFromKafka(ExternalTransform):
    schema = ReadFromKafkaSchema
    urn = 'beam:external:java:kafka:read:v1'
```

That's all the would be required from a user.  We would automatically handle converting `**kwargs` to a `ExternalConfigurationPayload` on `__init__`.

I like your suggestion of determining the coder automatically from the input type, but I think it might be too automagic.  An explicit schema provides more protection, and it's also a nice simple stand-in for a protobuf object.  I assume we don't want to use those here because it would require each new external transform to add to the beam model.

You may be wary of adding `attrs` as a dependency for this, but I've seen it used in a number of projects, since it answers such a fundamental problem in python.  It was the inspiration for the dataclass feature in python 3.7, but it continues to work on python2.  It also supports type analsys via mypy.   `attrs` is also an ideal library for creating data classes to use with beam transforms, since it's more structured than a dict, can enforce immutability, can be statically analyzed, and can be inspected to automatically serialize.  See [cattrs](https://cattrs.readthedocs.io/en/latest/) for an example of the latter.

Reply:
I like attrs, but I wouldn't want to force its use. The core value in this class is the mapping from dict + urn to an external transform application, and the users can decide how to define the dict.

Reply:
I've pushed what I think is a good compromise.  

Here's a brief example:

```python
class ReadFromKafka(External):
  byte_array_deserializer = 'org.apache.kafka.common.serialization.' \
                            'ByteArrayDeserializer'

  _urn = 'beam:external:java:kafka:read:v1'

  _schema = {
    'consumer_config': typehints.Iterable[typehints.KV[unicode, unicode]],
    'topics': typehints.Iterable[unicode],
    'key_deserializer': unicode,
    'value_deserializer': unicode,
  }

  def __init__(self, 
               consumer_config,
               topics,
               key_deserializer=byte_array_deserializer,
               value_deserializer=byte_array_deserializer,
               expansion_service=None):
    super(ReadFromKafka, self).__init__(expansion_service)
    self.consumer_config = list(consumer_config.items())
    self.topics = topics
    self.key_deserializer = key_deserializer
    self.value_deserializer = value_deserializer
```

The new `_schema` class attribute guides the extraction and encoding of the configuration payload.  

What I really like about this is that it sets us up for an even easier workflow when we go python3-only, in which case, we can drop the schema and instead read the typehints from the `__init__` annotations:

```python
class ReadFromKafka(External):
  byte_array_deserializer = 'org.apache.kafka.common.serialization.' \
                            'ByteArrayDeserializer'

  _urn = 'beam:external:java:kafka:read:v1'

  def __init__(self, 
               consumer_config: Iterable[Tuple[str, str]],
               topics: Iterable[str],
               key_deserializer: str = byte_array_deserializer,
               value_deserializer: str = byte_array_deserializer,
               expansion_service: Optional[str] = None):
    super(ReadFromKafka, self).__init__(expansion_service)
    self.consumer_config = list(consumer_config.items())
    self.topics = topics
    self.key_deserializer = key_deserializer
    self.value_deserializer = value_deserializer
```

It's still possible for classes to override various methods on the `External` base class to accomplish this however they like.  For example, if they want to do their own config extraction without relying on a schema, they can override `get_config()`, to control the encoding, they can override ``encode_config()` or if they want to do the whole thing, they can override `get_encoded_config()`. 

I provided an example of overriding `get_config()` on `GenerateSequence`.

--- 

There's one (known) problem left: now that it's inferring the coder from typehints I don't know when it's appropriate to automatically insert the `LengthPrefixCoder`.  Any suggestions?

Reply:
Oh, and by the way, in python3 , attrs copies the annotations into the `__init__` that it generates, so _if_ a user wanted to use attrs or dataclasses, this would also work (but would not be required):

```python
@attr.s
class ReadFromKafka(External):
  byte_array_deserializer = 'org.apache.kafka.common.serialization.' \
                            'ByteArrayDeserializer'

  _urn = 'beam:external:java:kafka:read:v1'

  consumer_config = attrib(type=Iterable[Tuple[str, str]])
  topics = attrib(type=Iterable[str])
  key_deserializer = attrib(type=str)
  value_deserializer = attrib(type=str)
  expansion_service = attrib(type=Optional[str])
```